packages/store-sync/src/watchLogs.ts

```diff
@@ -23,137 +26,227 @@ type WatchLogsEvent = {
           if (caughtUp) {
             subscriber.next({ blockNumber, logs: parsedLogs });
-            // Since this the event's block number corresponds to a pending block, we have to refetch this block in case of a restart
-            resumeBlock = blockNumber;
           } else {
             logBuffer.push(...parsedLogs);
           }
           return;
         }
+
+        debug(`ws${wsId} message`);
       });
 
       // Catch up to the pending logs
-      try {
-        const initialLogs = await fetchInitialLogs({ client, address, fromBlock: resumeBlock, topics });
-        debug("got", initialLogs.logs.length, "initial logs");
-        const logs = [...initialLogs.logs, ...logBuffer].sort(logSort);
-        debug("combining with buffer of", logBuffer.length, "logs");
-        const blockNumber = logs.at(-1)?.blockNumber ?? initialLogs.blockNumber;
-        subscriber.next({ blockNumber, logs });
-        // Since this the block number can correspond to a pending block, we have to refetch this block in case of a restart
-        resumeBlock = blockNumber;
-        caughtUp = true;
-      } catch (error) {
-        debug("could not get initial logs", error);
-        subscriber.error("Could not fetch initial wiresaw logs");
-      }
+      fetchInitialLogs({ ws, address, fromBlock, topics, wsId })
+        .then((initialLogs) => {
+          debug("got", initialLogs.logs.length, "initial logs");
+          const logs = [...initialLogs.logs, ...logBuffer].sort(logSort);
+          const blocks = groupLogsByBlockNumber(logs);
+          debug("releasing", logs.length, "logs across", blocks.length, "blocks to subscriber");
+          for (const block of blocks) {
+            subscriber.next(block);
+          }
+          caughtUp = true;
+          logBuffer.length = 0;
+        })
+        .catch((error) => {
+          const errorMessage = `ws${wsId} failed to fetch initial logs`;
+          debugError(errorMessage, error);
+          subscriber.error(errorMessage);
+        });
     }
 
-    setupClient().catch((error) => {
-      debug("error setting up initial client", error);
-      subscriber.error(error);
+    setup().catch((error) => {
+      const errorMessage = `ws${wsId} failed to setup wiresaw_watchLogs subscription`;
+      debugError(errorMessage, error);
+      subscriber.error(errorMessage);
     });
 
+    // Send a ping to keep the connection alive
+    const ping = setInterval(
+      () =>
+        ws &&
+        request({ ws, method: "net_version", wsId }).catch(() => {
+          const errorMessage = `ws${wsId} ping failed`;
+          debugError(errorMessage);
+          subscriber.error(errorMessage);
+        }),
+      10_000,
+    );
+
     return () => {
-      debug("logs$ subscription closed, closing client");
+      debug(`ws${wsId} logs$ subscription closed, closing client`);
+      clearInterval(ping);
       try {
-        client?.close();
-        client?.socket?.close();
+        ws?.close();
```

Comment:
doesn't look like there's anything here to auto-close after an error? intentional?

Reply:
yeah the idea is errors are passed to the subscriber to handle it, and stopping the subscription closes the websocket